thriftpy/parser/parser.py

```diff
@@ -880,3 +301,224 @@ def _get_ttype(inst, default_ttype=None):
+c_comment = '/*' <(~'*/' anything)*>:body '*/' -> body
+python_comment = '#' rest_of_line
+rest_of_line = <('\\\n' | (~'\n' anything))*>
+immutable = '(' brk 'python.immutable' brk '=' brk '""' brk ')'
+Reserved = ('__CLASS__' | '__DIR__' | '__FILE__' | '__FUNCTION__' | '__LINE__' | '__METHOD__' |
+            '__NAMESPACE__' | 'abstract' | 'alias' | 'and' | 'args' | 'as' | 'assert' | 'BEGIN' |
+            'begin' | 'binary' | 'bool' | 'break' | 'byte' | 'case' | 'catch' | 'class' | 'clone' |
+            'const' | 'continue' | 'declare' | 'def' | 'default' | 'del' | 'delete' | 'do' |
+            'double' | 'dynamic' | 'elif' | 'else' | 'elseif' | 'elsif' | 'END' | 'end' |
+            'enddeclare' | 'endfor' | 'endforeach' | 'endif' | 'endswitch' | 'endwhile' | 'ensure' |
+            'enum' | 'except' | 'exception' | 'exec' | 'extends' | 'finally' | 'float' | 'for' |
+            'foreach' | 'from' | 'function' | 'global' | 'goto' | 'i16' | 'i32' | 'i64' | 'if' |
+            'implements' | 'import' | 'in' | 'include' | 'inline' | 'instanceof' | 'interface' |
+            'is' | 'lambda' | 'list' | 'map' | 'module' | 'namespace' | 'native' | 'new' | 'next' |
+            'nil' | 'not' | 'oneway' | 'optional' | 'or' | 'pass' | 'print' | 'private' |
+            'protected' | 'public' | 'public' | 'raise' | 'redo' | 'register' | 'required' |
+            'rescue' | 'retry' | 'return' | 'self' | 'service' | 'set' | 'sizeof' | 'static' |
+            'string' | 'struct' | 'super' | 'switch' | 'synchronized' | 'then' | 'this' |
+            'throw' | 'throws' | 'transient' | 'try' | 'typedef' | 'undef' | 'union' | 'union' |
+            'unless' | 'unsigned' | 'until' | 'use' | 'var' | 'virtual' | 'void' | 'volatile' |
+            'when' | 'while' | 'with' | 'xor' | 'yield')
+not_reserved = ~(Reserved (' ' | '\t' | '\n'))
+'''
+
+
+BASE_TYPE_MAP = {
+    'bool': TType.BOOL,
+    'byte': TType.BYTE,
+    'i8': TType.BYTE,
+    'i16': TType.I16,
+    'i32': TType.I32,
+    'i64': TType.I64,
+    'double': TType.DOUBLE,
+    'string': TType.STRING,
+    'binary': TType.BINARY
+}
+
+
+PARSER = parsley.makeGrammar(
+    GRAMMAR, 
+    {
+        'Document': collections.namedtuple('Document', 'headers definitions'),
+        'Include': _add_include,
+        'Definition': _add_definition,
+        'Enum': _make_enum,
+        '_fill_in_struct': _fill_in_struct,
+        'Exception_': _make_exception,
+        'Service': _make_service,
+        'Function': collections.namedtuple('Function', 'name ttype fields oneway throws'),
+        'Field': collections.namedtuple('Field', 'id req ttype name default'),
+        'IdentifierRef': _lookup_symbol,
+        'BaseTType': BASE_TYPE_MAP.get,
+        'TType': TType,
+        '_cast_struct': _cast_struct,
+        'DeclareStruct': _make_empty_struct,
+        '_ref_type': _ref_type,
+        '_ref_val': _ref_val,
+        '_attr_ttype': _attr_ttype,
+    }
+)
+
+
+class ParseError(ThriftGrammerError): pass
```

Comment:
I think it might help if all exception types were grouped above, with docstrings about their usage.